### PR TITLE
http proxy: allow multiple listeners

### PR DESCRIPTION
**Pull request by mmatczuk**
Add HTTPProxyConfig.ExtraListeners and refactor code to work with multiple listeners.

### CODE REVIEW

File: http_proxy.go [hunk 1]
```diff
@@ -500,79 +504,116 @@ func (hp *HTTPProxy) handler() http.Handler {
...
+			hp.log.Errorf("failed to shutdown server error=%s", err)
 		}
-		return srvErr
+		return ctx.Err()
+	})
+	for i := range hp.listeners {
+		l := hp.listeners[i]
+		g.Go(func() error {
+			return srv.Serve(l)
+		})
```
Comment (Choraden) [inline http_proxy.go:535]:
```
^C2024/11/07 14:41:01.894475 [proxy] [INFO] closing down proxy
2024/11/07 14:41:01.894515 [proxy] [INFO] waiting for connections to close
2024/11/07 14:41:01.894522 [proxy] [INFO] all connections closed
2024/11/07 14:41:01.894584 [ERROR] fatal error exiting: http: Server closed
exit status 1
```
Closing a proxy in a handler mode results in a fatal error.

File: http_proxy.go [hunk 2]
```diff
@@ -500,79 +504,116 @@ func (hp *HTTPProxy) handler() http.Handler {
...
-	l := Listener{
-		ListenerConfig: hp.config.ListenerConfig,
-		TLSConfig:      hp.tlsConfig,
-		PromConfig: PromConfig{
-			PromNamespace: hp.config.PromNamespace,
-			PromRegistry:  hp.config.PromRegistry,
-		},
+	prototype := Listener{
+		TLSConfig: hp.tlsConfig,
+		metrics:   newListenerMetrics(hp.config.PromRegistry, hp.config.PromNamespace),
```
Comment (Choraden) [inline http_proxy.go:569]: How about adding a `name` to the ListenerConfig and using it in metrics? That way we could have them more detailed e.g. which listener accepts the most.

**Comment by Choraden** [comment]
LGTM, personally I would go with a wrapper over a list of listeners that returns any accepted connection from the underlying slice, but this is even better!

Left a proposition on adding "named" listeners.

**Comment by mmatczuk** [comment]
MultiListener is complicated its main disadvantage is that you accept a connection and then send it to worker. It's not guaranteed there is a worker available. It could be done but the standard interfaces are not well suited for that.

File: http_proxy.go (hunk 2 above)
Comment (mmatczuk) [reply http_proxy.go:569]: Sure it's a good idea.

**Comment by mmatczuk** [comment]
Having considered it I'd add static address label so that we can distinguish them

**Comment by mmatczuk** [comment]
@Choraden PTAL

**Comment by mmatczuk** [comment]
Squashed some commits and rebased.

**Review by Choraden — APPROVED** [review]

**Pull request merged**